src/solver/optimisation/AreaBalance.h

```diff
@@ -0,0 +1,25 @@
+#pragma once
+#include "new_constraint_builder.h"
+
+struct AreaBalanceData
+{
+    std::vector<int>& NumeroDeContrainteDesBilansPays;
+    std::vector<int>& InjectionVariable;
+    std::vector<int>& WithdrawalVariable;
+    const std::vector<int>& IndexDebutIntercoOrigine;
+    const std::vector<int>& IndexSuivantIntercoOrigine;
+    const std::vector<int>& IndexDebutIntercoExtremite;
+    const std::vector<int>& IndexSuivantIntercoExtremite;
+    const PALIERS_THERMIQUES& PaliersThermiquesDuPays;
+    const std::vector<::ShortTermStorage::AREA_INPUT>& ShortTermStorage;
+};
+
+class AreaBalance : public NewConstraint
+{
+public:
+    using NewConstraint::NewConstraint;
+
+    void add(int pdt, int pays, std::shared_ptr<AreaBalanceData> data);
```

Comment:
We must refrain, when not necessary, to use heap allocation (in particular for perf reason), and even more shared ownership (little perf but mostly for clarity sake).

Here the method will clearly not take ownership of the data, and not modify it --> it must be passed by const reference.
```cpp
void add(int pdt, int pays, const AreaBalanceData& data);
```
Same thing everywhere in other constraint types.

Reply:
i'v tried refs data & faced huge amount of segfault-ish bug #1617

Reply:
I'll try again

Reply:
modify just a little bit : `data->NumeroDeContrainteDesBilansPays[pays]=.....`

Reply:
Ok! maybe a non const ref is needed.

Ok for the segaults issues: well, I think if we come over that again, we can find out another solution than doing heap allocation, because this should not really change the lifetime of our objects here (not completely true, they might be moved and possibly copied in a few places, but I don't think it should be an issue). Easier said than done though :)